---
title:  "Classifying Transcripts using DNS"
author:  "Zhezhen Wang and Biniam Feleke"
date:  "`r format(Sys.Date(), '%m/%d/%Y')`"
abstract:  >
  The purpose of this mini package is to class both coding and noncoding
  transcripts into biotypes using the most recent GENCODE annotations. This tool
  can also be used to define potential lncRNAs, given an available genome
  transcriptome assembly (a gtf file) or any genomic loci of interest. Any long
  transcripts that cover more than two coding regions of a genome can be
  identified using this package.
output:  
  html_document: default
  pdf_document: default
  word_document: default
vignette: >
  %\VignetteIndexEntry{title}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---
###### ###### 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

  * ##### [Standard workflow](#Standard workflow)
    *  ##### [Quick Start](#Quick Start)
    *  ##### [Genomic Data Source](#Genomic Data Source)
    *  ##### [Extracting Summary Data](#Extracting Summary Data)
    *  ##### [Loading Data](#Loading Data)
    *  ##### [Filtering gencode_gr](#Filtering gencode_gr)
    *  ##### [Examples](#Examples)
  * ##### [Acknowledgements](#Acknowledgements)
  * ##### [SessionInfo](#SessionInfo)
  * ##### [References](#References)

<a name="Quick Start"></a>\

#### __Quick Start__

  The general analysis overview for our package is shown below. First load
  libraray("DNS") and system datasets. Then excute datasets using getBiotypes
  and getReadthrough functions included in DNS package.

```{r, eval=TRUE, message=TRUE, warning=TRUE}
library(DNS)
data("gencode_gr.v19_chr21")
head(gencode_gr)
```

<a name="Genomic Data Source"></a>\

#### __Genomic Data Source__ 

  A raw sequencing can be obtained from various sources. We obtained a
  comprehensive gene annotation reference of human GRCh37 from
  [GENCODE](https://www.gencodegenes.org/human).
  The __getBiotypes__ function requires specific input file structure similar to
  a standard general transfer format(gtf) file. The gtf includes rows and
  columns (fields). The number of rows depend on the number of reads. There are
  eight fixed columns with specific headers. For details of the gtf file format
  visit this
  [link](https://useast.ensembl.org/info/website/upload/gff.html#tracklines target="_blank").
  The image below illustrates a standard gtf file format.


   ![Chromosome 21 of human GRCh37 gtf](chr21.jpg)

<a name="">[GoTop](#)</a> 
<a name="Extracting Summary Data"></a>\

##### __Extracting Summary Data__


  There are two way to get genome biotypes: a) "transctipt_type" b) "gene_type".
  We used the "transcript_type" to extract the regions of interest using python
  as shown below. __Note__ that the *"PATH_FILE"* refers to the path of the
  downloded gtf file. Make sure to replac the *"PATH_FILE"*.
  
**Python codes:**

```{r "python code", eval = FALSE}
gtf = ('PATH_FILE')
outF = open('gtf_summary_transbiotype.txt','w')

def getquote(str,f,target):
    targetLen = len(target)+2
    strInd = str.find(target)
    st = strInd + len(target)+2
    ed = st + str[st:].find('";')
    #print(st,ed)
    f.write('\t'+str[st:ed]) if strInd!= -1 else f.write('\t'+'NA.')

with open(gtf, 'r') as f:
     for line in f:
        if line[0] != '#':
            chromosome = line.split('\t')[0]
            st = line.split('\t')[3]
            ed = line.split('\t')[4]
            strand = line.split('\t')[6]
            type = line.split('\t')[2]
            outF.write(chromosome+'\t'+st+'\t'+ed+'\t'+strand+'\t'+type)
            c = 'transcript_id'
            g = 'gene_name'
            t = 'transcript_type'
            getquote(line,outF,c)
            getquote(line,outF,g)
            getquote(line,outF,t)
            outF.write('\n')
outF.close() 
```
***
<a name="Home">[GoTop](#)</a>

<a name="Loading Data"></a>\ 

#### __Loading Data__


  To load data from local drive use the following format shown below. __Note__
  that the *"PATH_FILE"* refers to the location of the extracted summary of
  genomic data shown above. For more way of loading data click
  [here](https://support.rstudio.com/hc/en-us/articles/218611977-Importing-Data-with-RStudio).

     #load data from local drive
     data <- read.delim("PATH_FILE", comment.char = "#")

     #load data internal system eg. in our package "DNS"

```{r quickstart}
library(DNS)
data("gencode_gr.v19_chr21")
#head(gencod_gr)
```

<a name="Home">[GoTop](#)</a>

<a name="Filtering gencode_gr"></a>\ 

#### __Filtering *gencode_gr*__

  Here we show an extraction of 'gencode_gr' dataset using R commands shown
  below. Note to replace *"PATH_FILE"* with file direcotry path.

    > gtf = read.table('PATH_FILE')
    > gtf = subset(gtf, V5 == 'transcript')
    > colnames(gtf) = c('chr','start','end','strand','biotype')
    > gr = GRanges(gtf)

<a name="Home">[GoTop](#)</a> 
<a name="Examples"></a>\

#### __Examples__

##### 1) Processing Query
***

```{r}
query <- GRanges(c('chr1:2-10:+','chr1:6-10:-'),Row.names = c('trans1','trans2'),score = c(1,2))
head(query)

```

##### 2) Classifying Biotypes
***

```{r}
library(DNS)
gr <- GRanges(c('chr1:1-5:+','chr1:2-3:+'),biotype = c('lincRNA','CPC'))
head(gr)
```


##### 3) Extracting intron coordinates
***

      # Intron coordinates
      
       intron <- GRanges('chr1:6-8:+')
  
```{r}
library(DNS)
intron <- GRanges('chr1:6-8:+')
head(intron)
```


##### 4) Filtering coding transcripts
***

    # Filtering coding regions using products from example 1, 2 and 3

```{r}
library(DNS)
coding_trncp <- getBiotypes(query, gr, intron)
head(coding_trncp)
```

    # Coding transcripts from internal datasets

```{r}
library(DNS)
data("intron_gr.chr21")
data("ILEF_gr.chr21")
data("gencode_gr.v19_chr21")

biotyp <- getBiotypes(ILEF_gr, gencode_gr, intron_gr)
head(biotyp)
```


##### 5) Finding overlapping transcripts
***

    # Overlapping of two coding regions 
    
```{r DNS}
library(DNS)
data("ILEF_gr.chr21")
data("gencode_gr.v19_chr21")
cod_gr <- subset(gencode_gr, biotype == 'protein_coding')

rdthrough <- getReadthrough(ILEF_gr,cod_gr)
head(rdthrough)
```

<a name="Home">[GoTop](#)</a>
<a name="Acknowledgements"></a>\

### __Acknowledgements__

  The development of this mini-package will not be possible without continous
  help and feedback from individuals and institutions including:
  The Bioconductor Core Team, Dr. Xianan Yang ,Dr. Tzintzuni Garcia and and National Institutes of Health R21LM012619.

<a name="SessionInfo"></a>\

```{r SessionInfo}
sessionInfo()
```
<a name="References"></a>
<a name="Home">[GoTop](#)</a>\

### __References__ 

  * Wang, Z. Z., J. M. Cunningham and X. H. Yang (2018). "CisPi: a transcriptomic
score for disclosing cis-acting disease-associated lincRNAs."
Bioinformatics34(17): 664-670"

